src/main/java/crazypants/enderio/teleport/packet/PacketLongDistanceTravelEvent.java

```diff
@@ -69,51 +95,105 @@ public IMessage onMessage(PacketLongDistanceTravelEvent message, MessageContext
 
         TravelSource source = TravelSource.values()[message.source];
 
-        if (!validate(toTp, source)) {
+        if (!validate(
+                toTp,
+                source,
+                message.conserveMotion,
+                message.alsoDoTeleport,
+                message.tpX,
+                message.tpY,
+                message.tpZ)) {
             Log.LOGGER.warn(
                     Log.securityMarker,
                     "Player {} tried to tp without valid prereq.",
                     ctx.getServerHandler().playerEntity.getGameProfile());
             return null;
         }
 
-        doServerTeleport(toTp, message.conserveMotion, source);
+        doServerTeleport(
+                toTp,
+                message.conserveMotion,
+                source,
+                message.alsoDoTeleport,
+                message.tpX,
+                message.tpY,
+                message.tpZ);
 
         return null;
     }
 
-    private static boolean validate(EntityPlayerMP toTp, TravelSource source) {
+    private static boolean validate(EntityPlayerMP toTp, TravelSource source, boolean conserveMotion,
+            boolean alsoDoTeleport, int tpX, int tpY, int tpZ) {
         ItemStack equippedItem = toTp.getCurrentEquippedItem();
         switch (source) {
             case STAFF:
+                // Not allowed to do teleport staff teleport with regular staff.
+                if (alsoDoTeleport) {
+                    return false;
+                }
                 return equippedItem != null && equippedItem.getItem() instanceof IItemOfTravel
                         && ((IItemOfTravel) equippedItem.getItem()).isActive(toTp, equippedItem);
             case TELEPORT_STAFF:
                 // tp staff is creative version of traveling staff
                 // no energy check or anything else needed
                 // but the player must actually be equipped with one of these
+                if (alsoDoTeleport) {
+                    String error = TravelController.instance
+                            .validatePacketTravelEvent(toTp, tpX, tpY, tpZ, 0, conserveMotion, source);
+                    if (error != null) {
+                        return false;
+                    }
+                }
                 return equippedItem != null && equippedItem.getItem() instanceof ItemTeleportStaff;
+
             default:
                 // all other types are not allowed
                 return false;
         }
     }
 
-    public static boolean doServerTeleport(Entity toTp, boolean conserveMotion, TravelSource source) {
-        EntityPlayer player = toTp instanceof EntityPlayer ? (EntityPlayer) toTp : null;
+    public static boolean doServerTeleport(Entity toTp, boolean conserveMotion, TravelSource source,
+            boolean alsoDoTeleport, int tpX, int tpY, int tpZ) {
+        if (!(toTp instanceof EntityPlayer)) {
+            return false;
+        }
+        EntityPlayer player = (EntityPlayer) toTp;
```

Comment:
We actually validate on line 84 of this file that it's teleporting a player, so it is indeed impossible to use it to teleport non-players. This is also true of `PacketTravelEvent` (which this class was based off of).

You're right that it would probably be best to change the method signature, which would also allow removing the `entityId` field. I would prefer to do this in a separate PR, and refactor both packets at once, though, as this PR is probably too large and tangled already. Let me know if you really want me to do it in this PR, though (and if you want both packets refactored).